### PR TITLE
Arm build changes

### PR DESCRIPTION
**Pull request by michaelgsharp**
Changes that allow building for arm/arm64/apple silicon, as well as cross targeting to those same architectures.

Tests don't pass on those architectures yet, this is just enabling building. Anything that doesn't depend on x86/64 SIMD or IntelMKL works correctly.

### CODE REVIEW

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj [hunk 1]
```diff
@@ -16,7 +16,7 @@
     </ProjectReference>
 
     <NativeAssemblyReference Include="MatrixFactorizationNative" />
-    <NativeAssemblyReference Include="FastTreeNative" />
+    <NativeAssemblyReference Condition="'$(TargetArchitecture)' != 'arm64' And '$(TargetArchitecture)' != 'arm'" Include="FastTreeNative" />
```
Comment (eerhardt) [inline docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:19]: I wonder if we can create a helper property in Directory.Build.props `$(TargetsArm)` or similar. Then we don't need this duplicated everywhere.

File: eng/common/cross/build-rootfs.sh [hunk 2]
```diff
@@ -56,6 +56,10 @@ __UbuntuPackages+=" libkrb5-dev"
 __UbuntuPackages+=" libssl-dev"
 __UbuntuPackages+=" zlib1g-dev"
 
+# ML.NET dependencies
```
Comment (eerhardt) [inline eng/common/cross/build-rootfs.sh:59]: Since this is in `eng\common`, should this be changed in `dotnet/arcade`?

File: src/Microsoft.ML.Console/Microsoft.ML.Console.csproj [hunk 3]
```diff
@@ -28,13 +28,14 @@
...
-    <NativeAssemblyReference Include="FastTreeNative" />
-    <NativeAssemblyReference Include="CpuMathNative" />
+    <NativeAssemblyReference Condition="'$(TargetArchitecture)' != 'arm64' And '$(TargetArchitecture)' != 'arm'" Include="FastTreeNative" />
+    <NativeAssemblyReference Condition="'$(TargetArchitecture)' != 'arm64' And '$(TargetArchitecture)' != 'arm'" Include="CpuMathNative" />
     <NativeAssemblyReference Include="MatrixFactorizationNative" />
     <NativeAssemblyReference Include="LdaNative" />
-    <NativeAssemblyReference Include="SymSgdNative"/>
-    <NativeAssemblyReference Include="MklImports"/>
-    <NativeAssemblyReference Condition="'$(OS)' == 'Windows_NT'" Include="libiomp5md"/>
+    <!-- TODO: once we fix the 4 intel MKL methods, SymSgdNative will need to go back in. -->
```
Comment (eerhardt) [inline src/Microsoft.ML.Console/Microsoft.ML.Console.csproj:35]: How are we tracking this TODO?

File: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj [hunk 4]
```diff
@@ -4,7 +4,8 @@
     <TargetFramework>netstandard2.0</TargetFramework>
     <IncludeInPackage>Microsoft.ML.FastTree</IncludeInPackage>
     <PackageDescription>ML.NET component for FastTree</PackageDescription>
-    <DefineConstants>$(DefineConstants);USE_FASTTREENATIVE;NO_STORE;CORECLR</DefineConstants>
+    <DefineConstants>$(DefineConstants);NO_STORE;CORECLR</DefineConstants>
+    <DefineConstants Condition="'$(TargetArchitecture)' != 'arm64' And '$(TargetArchitecture)' != 'arm'">$(DefineConstants);USE_FASTTREENATIVE</DefineConstants>
```
Comment (eerhardt) [inline src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj:8]: I don't think this is going to work. We only put 1 managed Microsoft.ML.FastTree.dll in our NuGet package.

We will either need to build the C# code twice, and include each in a RID-specific section of the NuGet package.

Or we will need to continue building the C# code once, and then have a switch at runtime whether we are running on ARM or not.

File: src/Native/MatrixFactorizationNative/CMakeLists.txt [hunk 5]
```diff
@@ -1,16 +1,32 @@
...
+endif()
 
 include_directories(libmf)
 
 if(UNIX)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -O3 -pthread -std=c++0x -march=native")
+    if(NOT ${ARCHITECTURE} MATCHES "arm.*")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -O3 -pthread -std=c++0x -march=native")
+    else()
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -O3 -pthread -std=c++0x")
```
Comment (eerhardt) [inline src/Native/MatrixFactorizationNative/CMakeLists.txt:15]: Can we factor this so it isn't duplicated?

File: src/Native/build.sh [hunk 6]
```diff
@@ -103,6 +103,23 @@ fi
 
 __cmake_defines="${__cmake_defines} -DVERSION_FILE_PATH:STRING=${__versionSourceFile}"
 
+OS_ARCH=$(uname -m)
+OS=$(uname)
+
+# If we are cross compiling on Linux we need to set the CMAKE_TOOLCHAIN_FILE
+if [[ ( $OS_ARCH == "amd64" || $OS_ARCH == "x86_64" ) && ( $__build_arch == "arm64" || $__build_arch == "arm" ) && $OS != "Darwin" ]] ; then
+    __cmake_defines="${__cmake_defines} -DCMAKE_TOOLCHAIN_FILE=$(pwd)/../../eng/common/cross/toolchain.cmake"
```
Comment (eerhardt) [inline src/Native/build.sh:111]: You can use `$RootRepo/eng/common/...` here instead of assuming where this file exists in the source tree.

File: eng/common/cross/build-rootfs.sh (hunk 2 above)
Comment (michaelgsharp) [reply eng/common/cross/build-rootfs.sh:59]: I have already made this change in `dotnet/arcade` and the PR is merged, but the latest version of arcade breaks the ML.NET build. While I am troubleshooting the issue I have made the change locally so we can still take advantage of it.

File: src/Microsoft.ML.Console/Microsoft.ML.Console.csproj (hunk 3 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Console/Microsoft.ML.Console.csproj:35]: I have it tracked locally, but I will make an issue for it. There are 4 methods that deal with vector products that just need to be replaced so SymSgd can work on arm.

File: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj:8]: Good point. I think that having the runtime switch is better. Instead of checking if we are running on arm or not I'll just check if the native dll is present or not. That way even blazer would work with this.

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj (hunk 1 above)
Comment (michaelgsharp) [reply docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:19]: Yep, good idea. I'll go ahead and do that.

File: src/Native/MatrixFactorizationNative/CMakeLists.txt (hunk 5 above)
Comment (michaelgsharp) [reply src/Native/MatrixFactorizationNative/CMakeLists.txt:15]: Refactored.

File: src/Native/build.sh (hunk 6 above)
Comment (michaelgsharp) [reply src/Native/build.sh:111]: Thats good to know. I had tried `$RepoRoot` and that didn't work so I assumed there wasn't anything for it. Code has been changed.

File: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj]: Alright, I have added a runtime switch for this. It uses a delegate to pick between the native/managed version so it shouldn't have any performance impact, but if you could take a look I would appreciate it. I am gonig to run the benchmark tests before/after as well.

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj (hunk 1 above)
Comment (michaelgsharp) [reply docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj]: Ok I have added this. Take a look and see if it was what you were thinking. @safern can you take a look as well since you know msbuild really well? `Directory.Build.props` lines 30 - 37.

File: Directory.Build.props [hunk 7]
```diff
@@ -23,12 +23,21 @@
...
     <StrongNameKeyId>Open</StrongNameKeyId>
   </PropertyGroup>
 
+  <ItemGroup>
+    <TargetsArm Include="LdaNative" />
+    <TargetsArm Include="MatrixFactorizationNative" />
+  </ItemGroup>
+
+  <PropertyGroup>
+    <DefineConstants Condition="'$(TargetArchitecture)' != 'arm64' And '$(TargetArchitecture)' != 'arm'">$(DefineConstants);ARM64</DefineConstants>
```
Comment (eerhardt) [inline Directory.Build.props:38]: What is this DefineConstants for? Who uses it?

Comment (eerhardt) [reply Directory.Build.props:38]: Also, isn't this logic backwards?

File: src/Microsoft.ML.Console/Microsoft.ML.Console.csproj [hunk 8]
```diff
@@ -32,9 +32,10 @@
     <NativeAssemblyReference Include="CpuMathNative" />
     <NativeAssemblyReference Include="MatrixFactorizationNative" />
     <NativeAssemblyReference Include="LdaNative" />
+    <!-- TODO: once we fix the 4 intel MKL methods, SymSgdNative will need to go back in. -->
```
Comment (eerhardt) [inline src/Microsoft.ML.Console/Microsoft.ML.Console.csproj:35]: Can we remove this TODO now?

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs [hunk 9]
```diff
@@ -32,6 +34,38 @@ internal abstract class IntArray : IEnumerable<int>
...
+        /// or the managed fallbacks.
+        /// </summary>
+        public static Lazy<bool> UseFastTreeNative = new(() =>
+        {
+            try
+            {
+                NonExistantMethod();
+            }
+            // If the entry point isn't found then the dll exists and we should use the managed code.
+            catch(EntryPointNotFoundException)
```
Comment (eerhardt) [inline src/Microsoft.ML.FastTree/Dataset/IntArray.cs:55]: This seems like a hack. Can we just check the Current architecture, if it is x86 or x64, we call into the native code. Else, use the C# fallback.

[https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture)

File: test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj [hunk 10]
```diff
@@ -35,6 +35,7 @@
     <ProjectReference Include="..\Microsoft.ML.TestFramework\Microsoft.ML.TestFramework.csproj" />
   </ItemGroup>
 
+  <!-- TODO: once we fix the 4 intel MKL methods, SymSgdNative will need to go back in. -->
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj:38]: Is this TODO still necessary?

File: src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Microsoft.ML.FastTree.csproj]: So the benchmarks are basically identical speedwise before and after these changes. I had our benchmark run 30 times and the mean of the current code and the new code are within .15 seconds of each other on my local machine. So there shouldn't be noticeable performance impact from doing it this way.

File: Directory.Build.props (hunk 7 above)
Comment (michaelgsharp) [reply Directory.Build.props:38]: No one, I'll remove it. Its an accidental left over from something I was testing.

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs (hunk 9 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Dataset/IntArray.cs:55]: I did it this way for 2 reasons.

1 - So that the managed fallback would work even on an x86 or x64 machine if desired. Do you think that situation will happen?

2 - The docs you linked say that call is supported starting in .NET Framework 4.7.1, but ML.NET supports .NET Framework 4.6.1. I still get mixed up about which things work as long as you build with a later version and which things require the later version at runtime. Would this still work if it was built with a version that supports it but runs on a version that doesn't?

Comment (eerhardt) [reply src/Microsoft.ML.FastTree/Dataset/IntArray.cs:55]: > So that the managed fallback would work even on an x86 or x64 machine if desired. Do you think that situation will happen?

It hasn't happened yet, so I don't think it is necessarily something we need to worry about. We don't do it in any other places we call into native code.

> The docs you linked say that call is supported starting in .NET Framework 4.7.1

That is a bit mis-leading. What that information is trying to convey is that the API is "inbox" on .NET Framework 4.7.1. If you look at the next row down you'll see the API supports .NET Standard `1.1` - `2.1`. And since ML.NET targets `netstandard2.0`, we can use this API.

File: src/Microsoft.ML.Console/Microsoft.ML.Console.csproj (hunk 8 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Console/Microsoft.ML.Console.csproj:35]: Yes, removed.

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs (hunk 9 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Dataset/IntArray.cs:55]: Good to know. I have switched it to use the code you suggested.

File: test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj (hunk 10 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj:38]: Nope. Removed.

File: test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj [hunk 11]
```diff
@@ -19,13 +19,13 @@
     <PackageReference Include="Microsoft.Extensions.ObjectPool" Version="$(MicrosoftExtensionsTestPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.Options" Version="$(MicrosoftExtensionsTestPackageVersion)" />
   </ItemGroup>
-
```
Comment (eerhardt) [inline test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj:22]: Would it be possible to revert the unnecessary changes in this PR? If we want to make these formatting changes, I think they should be a separate PR.

File: test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj [hunk 12]
```diff
@@ -21,6 +21,7 @@
     </None>
   </ItemGroup>
 
+  <!-- TODO: once we fix the 4 intel MKL methods, SymSgdNative will need to go back in. -->
```
Comment (eerhardt) [inline test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj:24]: I still see these TODOs in a few places. 

* test/Microsoft.ML.IntegrationTests/Microsoft.ML.IntegrationTests.csproj
* test/Microsoft.ML.NightlyBuild.Tests/Microsoft.ML.NightlyBuild.Tests.csproj
* test/Microsoft.ML.Predictor.Tests/Microsoft.ML.Predictor.Tests.csproj

Should they all be removed?

File: Directory.Build.targets [hunk 13]
```diff
@@ -8,28 +8,34 @@
...
-			<FullAssemblyPath>$(NativeOutputPath)$(LibPrefix)%(NativeAssemblyReference.Identity)$(LibExtension)</FullAssemblyPath>
-		</NativeAssemblyReference>
+      <NativeAssemblyReference>
+        <FullAssemblyPath>$(NativeOutputPath)$(LibPrefix)%(NativeAssemblyReference.Identity)$(LibExtension)</FullAssemblyPath>
+      </NativeAssemblyReference>
 		</ItemGroup>
 
-		<Copy SourceFiles = "@(NativeAssemblyReference->'%(FullAssemblyPath)')"
+    <PropertyGroup>
+      <ShouldCopyx64>false</ShouldCopyx64>
```
Comment (eerhardt) [inline Directory.Build.targets:31]: I'm not sure I understand the value of these properties. What other target architectures do we have beside x86, x64, arm64, or arm?

These properties appear to only be used in the below Copy, and as far as I can tell, that condition is always true.

Comment (eerhardt) [reply Directory.Build.targets:31]: Oh, I see now, this is to support copying `LdaNative` and `MatrixFactorizationNative` assemblies on arm, but not the other native files.

File: Directory.Build.props [hunk 14]
```diff
@@ -23,12 +23,17 @@
...
-    <TargetArchitecture Condition="'$(TargetArchitecture)' == ''">x64</TargetArchitecture>
-    <NativeTargetArchitecture Condition="'$(NativeTargetArchitecture)' == ''">$(TargetArchitecture)</NativeTargetArchitecture>
+    <TargetArchitecture Condition="'$(TargetArchitecture)' == ''">$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture)</TargetArchitecture>
+    <TargetArchitecture>$(TargetArchitecture.ToLower())</TargetArchitecture>
     <PlatformConfig>$(Platform).$(Configuration)</PlatformConfig>
     <StrongNameKeyId>Open</StrongNameKeyId>
   </PropertyGroup>
 
+  <ItemGroup>
+    <TargetsArm Include="LdaNative" />
```
Comment (eerhardt) [inline Directory.Build.props:33]: This item name is a bit odd. Maybe it should be `NativeAssembliesTargettingArm` or similar?

Comment (eerhardt) [reply Directory.Build.props:33]: We should also consider putting this next to the place in the `.targets` file that uses it. No real need to separate this into the `.props` file, when only 1 place in a target reads it.

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs [hunk 15]
```diff
@@ -32,6 +34,25 @@ internal abstract class IntArray : IEnumerable<int>
         /// </summary>
         public abstract int Length { get; }
 
+        // We can test if the dll is present by calling into a non-existant method.
+        // If we get an EntryPointNotFoundException then the dll exists and we can use the native code.
+        // If we get a DllNotFoundException then the dll doesn't exist and we should use the managed fallbacks.
+        private const string NativePath = "FastTreeNative";
+        [DllImport(NativePath), SuppressUnmanagedCodeSecurity]
+        private static extern void NonExistantMethod();
```
Comment (eerhardt) [inline src/Microsoft.ML.FastTree/Dataset/IntArray.cs:42]: Is this still used?

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs [hunk 16]
```diff
@@ -32,6 +34,25 @@ internal abstract class IntArray : IEnumerable<int>
...
+
+        /// <summary>
+        /// Lazy bool that checks if we are on x86/x64 so we know if we should use the native code
+        /// or the managed fallbacks.
+        /// </summary>
+        public static Lazy<bool> UseFastTreeNative = new(() =>
```
Comment (eerhardt) [inline src/Microsoft.ML.FastTree/Dataset/IntArray.cs:48]: This doesn't need to be lazy. It can just be

```c#
public static bool UseFastTreeNative => RuntimeInformation.ProcessArchitecture == Architecture.X64 || RuntimeInformation.ProcessArchitecture == Architecture.X86
```

File: Directory.Build.props (hunk 14 above)
Comment (michaelgsharp) [reply Directory.Build.props:33]: Makes sense. Renamed and moved to the `.targets` file.

File: Directory.Build.targets (hunk 13 above)
Comment (michaelgsharp) [reply Directory.Build.targets:31]: Correct. I figured we could also use this same pattern when we get to blazer WASM. Though I guess we can't really copy any native files there... so maybe this was a bit overkill.

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs (hunk 15 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Dataset/IntArray.cs:42]: Nope, good catch. It has been removed.

File: src/Microsoft.ML.FastTree/Dataset/IntArray.cs (hunk 16 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.FastTree/Dataset/IntArray.cs:48]: Sounds good. Has been switched.

File: test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj (hunk 11 above)
Comment (michaelgsharp) [reply test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj:22]: Yep. I'll go ahead and revert them.

File: test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj (hunk 12 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj]: Yes, and they should be now.

File: test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj (hunk 11 above)
Comment (michaelgsharp) [reply test/Microsoft.Extensions.ML.Tests/Microsoft.Extensions.ML.Tests.csproj]: Reverted

File: src/Native/gen-buildsys-win.bat [hunk 17]
```diff
@@ -28,7 +28,9 @@ popd
 :DoGen
 if /i "%3" == "x64"     (set __ExtraCmakeParams=%__ExtraCmakeParams% -A x64)
 if /i "%3" == "x86"     (set __ExtraCmakeParams=%__ExtraCmakeParams% -A Win32)
-"%CMakePath%" "-DCMAKE_BUILD_TYPE=%CMAKE_BUILD_TYPE%" "-DCMAKE_INSTALL_PREFIX=%__CMakeBinDir%" "-DMKL_LIB_PATH=%MKL_LIB_PATH%" -G "Visual Studio %__VSString%" %__ExtraCmakeParams% -B. -H%1
+if /i "%3" == "arm64"     (set __ExtraCmakeParams=%__ExtraCmakeParams% -A arm64)
+if /i "%3" == "arm"     (set __ExtraCmakeParams=%__ExtraCmakeParams% -A arm)
+"%CMakePath%" "-DCMAKE_BUILD_TYPE=%CMAKE_BUILD_TYPE%" "-DCMAKE_INSTALL_PREFIX=%__CMakeBinDir%" "-DMKL_LIB_PATH=%MKL_LIB_PATH%" "-DARCHITECTURE=%3" -G "Visual Studio %__VSString%" %__ExtraCmakeParams% -B. -H%1
```
Comment (eerhardt) [inline src/Native/gen-buildsys-win.bat:33]: Why do we need both `-DARCHITECTURE` and `-A` above? Can we just use 1 command line arg to set the archiecture?

**Review by eerhardt — APPROVED** [review]
Just one minor question remaining. After answering that, this LGTM.

Nice work here!

File: src/Native/gen-buildsys-win.bat (hunk 17 above)
Comment (michaelgsharp) [reply src/Native/gen-buildsys-win.bat:33]: So the `-A` doesn't work on the Unix Makefile Generator, but that is how you set the architecture correctly for visual studio. I needed a way in the cmakelists file to be able to exclude native projects for both generators, and thats how I ended up with the `-DARCHITECTURE`. I'm sure there is a way I could use only the `-A` in visual studio and not need the `-DARCHITECTURE`, I just haven't been able to figure it out yet. We will always need `-DARCHITECTURE` for the Unix generator, its just if we can remove it from the visual studio one.

Comment (eerhardt) [reply src/Native/gen-buildsys-win.bat:33]: Have you looked how it is done in other repos? For example `dotnet/runtime`?

Comment (michaelgsharp) [reply src/Native/gen-buildsys-win.bat:33]: I'm checking that out right now. They pass the `-arch` flag. Thats a custom flag. I'm still investigating how they plumb it internally.

Comment (michaelgsharp) [reply src/Native/gen-buildsys-win.bat:33]: That eventually turns into this `/p:TargetArchitecture=$arch`. Still looking into how thats passed to the native side of things.

Comment (michaelgsharp) [reply src/Native/gen-buildsys-win.bat:33]: So it actually looks like they do the same thing. 
https://github.com/dotnet/runtime/blob/e4b4807e2fae2164d9116fbcdd49ba9044461e7e/eng/native/gen-buildsys.cmd#L34 they set the `-A` same way I do.

https://github.com/dotnet/runtime/blob/e4b4807e2fae2164d9116fbcdd49ba9044461e7e/src/coreclr/build-runtime.cmd#L450 they a flag `-DCLR_CMAKE_TARGET_ARCH` which is used the same why I am using the `ARCHITECTURE` flag. They do more complex stuff with it than we do since they target so many things, but its the same idea.

Comment (eerhardt) [reply src/Native/gen-buildsys-win.bat:33]: Sounds good. Thanks for verifying.
